### PR TITLE
[Metadata Management] added aws storage driver for metadata sync service

### PR DESCRIPTION
**Pull request by devanshjain7**
**What this PR does / why we need it**:
added aws storage driver for metadata sync service

**Test Report Added?**:
/kind TESTED

**Test Report**:
Metadata collected from aws s3 and synced with mongoDB: 
image

### CODE REVIEW

File: metadata/pkg/service/service.go [hunk 1]
```diff
@@ -23,7 +23,7 @@ import (
 
 	"github.com/aws/aws-sdk-go/aws/credentials"
 	backend "github.com/opensds/multi-cloud/backend/proto"
-	"github.com/opensds/multi-cloud/metadata/pkg/db"
+	// "github.com/opensds/multi-cloud/metadata/pkg/db"
```
Comment (josephjacobmorris) [inline metadata/pkg/service/service.go:26]: remove commented code

File: metadata/pkg/drivers/aws/aws.go [hunk 2]
```diff
@@ -43,10 +49,99 @@ type AwsAdapter struct {
...
+	}
+	wg.Wait()
+	return bucketArray, err
+}
+
 func (ad *AwsAdapter) SyncMetadata(ctx context.Context, in *pb.SyncMetadataRequest) error {
-	log.Infoln("Implement me")
+
+	buckArr, _ := BucketList(ad)
+	// out, _ := json.MarshalIndent(buckArr, "", "   ")
```
Comment (josephjacobmorris) [inline metadata/pkg/drivers/aws/aws.go:140]: remove commented code from this file too

**Comment by josephjacobmorris** [comment]
image
why two ids?

File: metadata/pkg/drivers/aws/aws.go [hunk 3]
```diff
@@ -43,10 +48,98 @@ type AwsAdapter struct {
 	Session *session.Session
 }
 
+func ObjectList(sess *session.Session, bucket *model.MetaBucket) error {
```
Comment (PravinRanjan10) [inline metadata/pkg/drivers/aws/aws.go:51]: please make method of aws type

File: metadata/pkg/drivers/aws/aws.go [hunk 4]
```diff
@@ -43,10 +48,98 @@ type AwsAdapter struct {
...
+		}(i, object)
+	}
+	wg.Wait()
+	bucket.NumberOfObjects = numObjects
+	bucket.TotalSize = totSize
+	bucket.Objects = objectArray
+	return err
+}
+
+func BucketList(sess *session.Session) ([]model.MetaBucket, error) {
```
Comment (PravinRanjan10) [inline metadata/pkg/drivers/aws/aws.go:98]: please make method of aws type

File: metadata/pkg/drivers/aws/aws.go [hunk 5]
```diff
@@ -43,10 +48,98 @@ type AwsAdapter struct {
...
+		go func(i int, object *s3.Object) {
+			defer wg.Done()
+			obj := model.MetaObject{}
+			obj.LastModifiedDate = *object.LastModified
+			obj.ObjectName = *object.Key
+			obj.Size = *object.Size
+			totSize += obj.Size
+			obj.StorageClass = *object.StorageClass
+
+			meta, _ := svc.HeadObject(&s3.HeadObjectInput{Bucket: &bucket.Name, Key: object.Key})
```
Comment (PravinRanjan10) [inline metadata/pkg/drivers/aws/aws.go:74]: handle the error

File: metadata/pkg/drivers/aws/aws.go [hunk 6]
```diff
@@ -43,10 +48,98 @@ type AwsAdapter struct {
...
+	}
+
+	numObjects := len(output.Contents)
+
+	objectArray := make([]model.MetaObject, numObjects)
+	wg := sync.WaitGroup{}
+	var totSize int64 = 0
+	for i, object := range output.Contents {
+		wg.Add(1)
+		go func(i int, object *s3.Object) {
```
Comment (PravinRanjan10) [inline metadata/pkg/drivers/aws/aws.go:65]: better to move in a separate function

File: metadata/pkg/drivers/aws/aws.go [hunk 7]
```diff
@@ -43,10 +48,98 @@ type AwsAdapter struct {
...
+	bucketArray := make([]model.MetaBucket, numBuckets)
+	wg := sync.WaitGroup{}
+	for i, bucket := range output.Buckets {
+		wg.Add(1)
+		go func(i int, bucket *s3.Bucket) {
+			defer wg.Done()
+			buck := model.MetaBucket{}
+			buck.CreationDate = *bucket.CreationDate
+			buck.Name = *bucket.Name
+			loc, _ := svc.GetBucketLocation(&s3.GetBucketLocationInput{Bucket: bucket.Name})
```
Comment (PravinRanjan10) [inline metadata/pkg/drivers/aws/aws.go:115]: handle the error

File: metadata/pkg/drivers/aws/aws.go [hunk 8]
```diff
@@ -43,10 +48,98 @@ type AwsAdapter struct {
...
+			buck.Region = *loc.LocationConstraint
+			tags, _ := svc.GetBucketTagging(&s3.GetBucketTaggingInput{Bucket: bucket.Name})
```
Comment (PravinRanjan10) [inline metadata/pkg/drivers/aws/aws.go:117]: handle the error

File: metadata/pkg/drivers/aws/aws.go [hunk 9]
```diff
@@ -43,10 +48,98 @@ type AwsAdapter struct {
...
+		}(i, bucket)
+	}
+	wg.Wait()
+	return bucketArray, err
+}
+
 func (ad *AwsAdapter) SyncMetadata(ctx context.Context, in *pb.SyncMetadataRequest) error {
-	log.Infoln("Implement me")
+
+	buckArr, _ := BucketList(ad.Session)
```
Comment (PravinRanjan10) [inline metadata/pkg/drivers/aws/aws.go:138]: handle the error

File: metadata/pkg/service/service.go [hunk 10]
```diff
@@ -23,7 +23,7 @@ import (
```
Comment (PravinRanjan10) [inline metadata/pkg/service/service.go:23]: why do you need this file, if there is no update??

**Review by PravinRanjan10 — CHANGES_REQUESTED** [review]

File: metadata/pkg/drivers/aws/aws.go (hunk 3 above)
Comment (devanshjain7) [reply metadata/pkg/drivers/aws/aws.go]: done

File: metadata/pkg/drivers/aws/aws.go (hunk 4 above)
Comment (devanshjain7) [reply metadata/pkg/drivers/aws/aws.go]: done

File: metadata/pkg/drivers/aws/aws.go (hunk 5 above)
Comment (devanshjain7) [reply metadata/pkg/drivers/aws/aws.go]: done

File: metadata/pkg/drivers/aws/aws.go (hunk 6 above)
Comment (devanshjain7) [reply metadata/pkg/drivers/aws/aws.go]: done

File: metadata/pkg/drivers/aws/aws.go (hunk 7 above)
Comment (devanshjain7) [reply metadata/pkg/drivers/aws/aws.go]: done

File: metadata/pkg/drivers/aws/aws.go (hunk 9 above)
Comment (devanshjain7) [reply metadata/pkg/drivers/aws/aws.go]: done

File: metadata/pkg/service/service.go (hunk 10 above)
Comment (devanshjain7) [reply metadata/pkg/service/service.go:23]: the metabucket model (from proto package) used in this file is different from what we are currently using for metadata collection. This causes errors while compiling code for metadata service. Hence, that part is commented for now.

File: metadata/pkg/drivers/aws/aws.go (hunk 8 above)
Comment (devanshjain7) [reply metadata/pkg/drivers/aws/aws.go]: done

File: metadata/pkg/db/drivers/mongo/mongo.go [hunk 11]
```diff
@@ -19,14 +19,14 @@ import (
...
-	"github.com/globalsign/mgo/bson"
+	"go.mongodb.org/mongo-driver/bson"
 	"github.com/micro/go-micro/v2/metadata"
 	log "github.com/sirupsen/logrus"
 	bson2 "go.mongodb.org/mongo-driver/bson"
 	"go.mongodb.org/mongo-driver/mongo"
 	"go.mongodb.org/mongo-driver/mongo/options"
 	"go.mongodb.org/mongo-driver/mongo/readpref"
-
+	backendpb "github.com/opensds/multi-cloud/backend/proto"
```
Comment (PravinRanjan10) [inline metadata/pkg/db/drivers/mongo/mongo.go:29]: please format it (probably remove extra space)

File: metadata/pkg/service/service.go (hunk 10 above)
Comment (devanshjain7) [reply metadata/pkg/service/service.go]: this issue is resolved and this file is unchanged.

File: metadata/pkg/drivers/aws/aws.go [hunk 12]
```diff
@@ -43,10 +50,122 @@ type AwsAdapter struct {
...
+	if err != nil {
+		log.Errorf("unable to list buckets. failed with error: %v\n", err)
+		return nil, err
+	}
+	numBuckets := len(output.Buckets)
+	bucketArray := make([]model.MetaBucket, numBuckets)
+	wg := sync.WaitGroup{}
+	for i, bucket := range output.Buckets {
+		wg.Add(1)
+		go iterateBuckets(i, bucket, sess, bucketArray, &wg)
```
Comment (PravinRanjan10) [inline metadata/pkg/drivers/aws/aws.go:152]: please don't use go routine here.(mostly it's nested go routines usage.)

File: metadata/pkg/drivers/aws/aws.go [hunk 13]
```diff
@@ -43,10 +50,122 @@ type AwsAdapter struct {
 	Session *session.Session
 }
 
+func iterateObjects(i int, object *s3.Object, svc *s3.S3, bucketName string, objectArray []model.MetaObject, wg *sync.WaitGroup) {
```
Comment (PravinRanjan10) [inline metadata/pkg/drivers/aws/aws.go:53]: IMO, GetObjectsParams or similar would be suitable name. You may consider changing the name

File: metadata/pkg/drivers/aws/aws.go [hunk 14]
```diff
@@ -43,10 +50,122 @@ type AwsAdapter struct {
...
+	defer wg.Done()
+	obj := model.MetaObject{}
+	obj.LastModifiedDate = *object.LastModified
```
Comment (PravinRanjan10) [inline metadata/pkg/drivers/aws/aws.go:56]: from line 56-59, can be used to fillied the model in previous function itself

File: metadata/pkg/drivers/aws/aws.go [hunk 15]
```diff
@@ -43,10 +50,122 @@ type AwsAdapter struct {
...
+	buck.CreationDate = *bucket.CreationDate
+	buck.Name = *bucket.Name
+	svc := s3.New(sess)
+	loc, err := svc.GetBucketLocation(&s3.GetBucketLocationInput{Bucket: bucket.Name})
+	if err != nil {
+		log.Errorf("unable to get bucket location. failed with error: %v\n", err)
+	} else {
+		buck.Region = *loc.LocationConstraint
+	}
+	newSvc := s3.New(sess, aws.NewConfig().WithRegion(buck.Region))
```
Comment (PravinRanjan10) [inline metadata/pkg/drivers/aws/aws.go:120]: you may ustilize previous session with region. Please check

**Review by PravinRanjan10 — CHANGES_REQUESTED** [review]

File: metadata/pkg/drivers/aws/aws.go (hunk 15 above)
Comment (devanshjain7) [reply metadata/pkg/drivers/aws/aws.go]: done

File: metadata/pkg/drivers/aws/aws.go (hunk 13 above)
Comment (devanshjain7) [reply metadata/pkg/drivers/aws/aws.go]: removed go routine at object level. so this function is removed.

File: metadata/pkg/drivers/aws/aws.go (hunk 12 above)
Comment (devanshjain7) [reply metadata/pkg/drivers/aws/aws.go]: done

File: metadata/pkg/drivers/aws/aws.go [hunk 16]
```diff
@@ -43,10 +50,122 @@ type AwsAdapter struct {
...
+func iterateBuckets(i int, bucket *s3.Bucket, sess *session.Session, bucketArray []model.MetaBucket, wg *sync.WaitGroup) {
+	defer wg.Done()
+	buck := model.MetaBucket{}
+	buck.CreationDate = *bucket.CreationDate
+	buck.Name = *bucket.Name
+	svc := s3.New(sess)
+	loc, err := svc.GetBucketLocation(&s3.GetBucketLocationInput{Bucket: bucket.Name})
+	if err != nil {
+		log.Errorf("unable to get bucket location. failed with error: %v\n", err)
+	} else {
```
[the remark this answers is not in the document]
Comment (devanshjain7) [reply metadata/pkg/drivers/aws/aws.go]: done

**Review by PravinRanjan10 — APPROVED** [review]
LGTM

**Pull request merged**